### PR TITLE
WIP: Fix compiler warnings

### PR DESCRIPTION
**Pull request by jmlemetayer**
Trying to fix all warnings noticed by #402 for the different builds:

- [x] linux / gcc / autotools
- [x] linux / clang / autotools
- [x] linux / gcc-5 / cmake
- [x] linux / clang-3.8 / cmake
- [x] osx / clang / cmake
- [x] osx / clang / autotools

### CODE REVIEW

**Comment by xor-gate** [comment]
looks good, but if we are going to move to a enforced codestyle with tooling (checkpatch + linux codestyle) typedefs are forbidden so we should change this to something like: `struct stlink_reg`.

**Comment by jmlemetayer** [comment]
I just wanted not to break the "code spirit". But if it's ok for you I will change this :smiley:

I have set this pull request as WIP to check the travis builds and also to have feedbacks. Thanks for your support!

**Comment by xor-gate** [comment]
Yeah I know what you mean, but in the upcoming time -> Milestone 2.0.0 there will some cleanup going on for easier understanding of the codebase and contribution.

**Comment by jmlemetayer** [comment]
Sounds interresting. If I have extra time I will think about it.

**Comment by xor-gate** [comment]
See https://github.com/texane/stlink/milestones -> 2.0.0

**Comment by xor-gate** [comment]
@jmlemetayer make sure you rebase with latest master because i have moved the flash loader functions into a seperate file.

**Comment by jmlemetayer** [comment]
The remainings warnings `typedef ‘_GStaticAssertCompileTimeAssertion_0’ locally defined but not used` are a known glib bug which has been fixed (I don't have this issue on my ubuntu 16.04).

@xor-gate Is that a problem for you?

**Comment by xor-gate** [comment]
We have to see how the glibc compiler warning will behave because I want to enable `-Werror` then all warnings are upgraded as errors. As for travis we run ubuntu 14 indeed. But there is a way to have newer VMs.

File: src/tools/gui/stlink-gui.c [hunk 1]
```diff
@@ -518,7 +516,7 @@ stlink_gui_set_connected (STlinkGUI *gui)
 }
 
 static void
-connect_button_cb (GtkWidget *widget, gpointer data)
+connect_button_cb (GtkWidget * UNUSED(widget), gpointer data)
```
Comment (xor-gate) [inline src/tools/gui/stlink-gui.c]: This is way ugly. You can just do it in a more portable way (pseudo-C):

<function> (param1, param2) {
     (void)param1;
     param2++;
}

**Comment by jmlemetayer** [comment]
Yeah, this is what I expected. I see 2 solutions:

 - Upgrade the GLib version in travis
 - Do not build GUI in travis (as the GLib is only used by GTK+ in the GUI)

File: include/stlink.h [hunk 2]
```diff
@@ -15,6 +15,15 @@
...
 
+#ifdef UNUSED
+#elif defined(__GNUC__)
+#define UNUSED(x)	UNUSED_ ## x __attribute__((unused))
+#elif defined(__LCLINT__)
+#define UNUSED(x)	/*@unused@*/ x
+#else
+#define UNUSED(x)	x
+#endif
+
```
Comment (xor-gate) [inline include/stlink.h]: We should not do this, see my other comment. As this is highly compiler depended and I want in some time also support native visual studio compilation. Also when we really need to do this, sure not in the public scope.

File: src/common.c [hunk 3]
```diff
@@ -181,6 +182,7 @@ static inline uint32_t read_flash_obr(stlink_t *sl) {
     stlink_read_debug32(sl, FLASH_OBR, &obr);
     return obr;
 }
+#endif
```
Comment (xor-gate) [inline src/common.c]: We need to throw this away when unused.

File: src/common.c [hunk 4]
```diff
@@ -366,11 +368,13 @@ static void set_flash_cr_strt(stlink_t *sl) {
     stlink_write_debug32(sl, cr_reg, val);
 }
 
+#if 0
 static inline uint32_t read_flash_acr(stlink_t *sl) {
     uint32_t acr;
     stlink_read_debug32(sl, FLASH_ACR, &acr);
     return acr;
 }
+#endif
```
Comment (xor-gate) [inline src/common.c]: Same comment as above, throw away when unused/dead code.

File: src/tools/flash.c [hunk 5]
```diff
@@ -84,18 +84,18 @@ static int get_opts(struct opts* o, int ac, char** av)
     {
         ac--;
         av++;
-	    int i=strlen(av[0]);
-	    if(i%2 != 0){
+	    int j = (int) strlen(av[0]);
```
Comment (xor-gate) [inline src/tools/flash.c]: this is not really portable, as strlen really returns `size_t` we need to obey and not cast it to a signed type.

File: src/tools/flash.c [hunk 6]
```diff
@@ -154,7 +154,7 @@ static int get_opts(struct opts* o, int ac, char** av)
     }
 
     o->filename = av[i + 1];
-    o->addr = strtoul(av[i + 2], NULL, 16);
+    o->addr = (uint32_t) strtoul(av[i + 2], NULL, 16);
```
Comment (xor-gate) [inline src/tools/flash.c]: this is a little evil as strtoul could return 0 and is of type `unsigned long int`

File: src/tools/term.c [hunk 7]
```diff
@@ -41,8 +41,8 @@ struct stlinky {
                                           | buf[2] << 16   \
                                           | buf[3] << 24))
 
-static stlink_t* sl;
-sigset_t sig_mask;
+static stlink_t* gsl;
```
Comment (xor-gate) [inline src/tools/term.c]: I'm not happy with a global, but prefixing it with g is not bad.

File: src/usb.c [hunk 8]
```diff
@@ -737,7 +737,7 @@ stlink_t *stlink_open_usb(enum ugly_loglevel verbose, bool reset, char serial[16
     }
 
     libusb_device **list;
-    int cnt = libusb_get_device_list(slu->libusb_ctx, &list);
+    int cnt = (int) libusb_get_device_list(slu->libusb_ctx, &list);
```
Comment (xor-gate) [inline src/usb.c]: We should use `ssize_t` and use it as a counter if > 0. As per libusb API:
`ssize_t 	libusb_get_device_list (libusb_context *ctx, libusb_device ***list)`

**Comment by xor-gate** [comment]
To make sure people dont break things we need to build the GUI so disable is no option. Upgrading, I have no clue and needs investigation (but what about people then building with older glibc will have problems). Have you tried the cmake `SYSTEM` directive?

Most of the refactoring seems fine, but a few `brutal-casts` need to be looked at better because the compiler is actually right about the problem. Of course a `brute-force` cast then the warning will go away but some bugs could come in.

**Comment by xor-gate** [comment]
Weird, no CI build is triggerd with last cmake fixup. Could you try to squash merge and push and see if it builds?

**Comment by jmlemetayer** [comment]
GitHub had some issue this afternoon. I will try again tomorrow. I think you can also trigger a new build directly from travis-CI.

**Comment by xor-gate** [comment]
Yeah I noticed also some bugging, you can only rebuild pushed commits. Thank you for you time and effort!

**Comment by jmlemetayer** [comment]
I remove the `UNUSED` macro as requested. I was not aware that the `(void)` trick is much more portable than a macro. I will check your other comments later.

I have added the `-Werror` and the `SYSTEM` but it doesn't seems to work with the `cmake 3.8` / `clang` couple...

**Comment by jmlemetayer** [comment]
@xor-gate About your comments above. I agree that it is not beautiful, but the code was already here.

For now, I don't think we should clean the code. Just fix warnings. The cleanup should be done in a second time.

**Comment by xor-gate** [comment]
You are right about that, maybe we should add doxygen style comments to track todos where it needs attention/refactoring for the second step like this:

```
/** @todo refactor/review */
```

Then when I enable building doxygen documentation we can see a list of todos or grep trough the code for todo. See https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdtodo

**Comment by xor-gate** [comment]
Could you try the following with the BEFORE statement, and see if it works?:

```cmake
include_directories(BEFORE SYSTEM <gtk include var>)
```

**Comment by xor-gate** [comment]
Hi @jmlemetayer what do you think, should we merge without `-Werror` because if this oddness?

**Comment by jmlemetayer** [comment]
Yes I think it is a good idea to remove `-Werror` for now. I did not manage to find a solution for the GLib issue.

**Comment by jmlemetayer** [comment]
From http://llvm.org/apt/

> APT mirror was temporary switched off due to excess load. We are working on bringing it back. Stay tuned!

All linux build are failed ...

**Comment by xor-gate** [comment]
Seems good to me, the llvm apt mirror is a different problem. We still need to refactor bits-and-pieces before hitting `v2.0.0`. Thank you for your work.

**Pull request merged**